File: javascript/blocks.js
/** Create a range of integers. */
function range(start, end) {
	var result = [];
	for ( var i = start; i < end; i++) {
		result.push(i);
	}
	return result;
}

/** Pick a random element from an array. */
function pickRandom(arr) {
	return arr[Math.floor(Math.random() * arr.length)];
}

/**  Block object */
function Block(x, y, color) {
	this.x = x;
	this.y = y;
	this.color = color;

	/** Create a new Block moved from this block. */
	this.move = function(dx, dy) {
		return new Block(x + dx, y + dy, this.color);
	}

	/** Rotate this block clockwise around a point. */
	this.rotateClockwise = function(pX, pY) {
		return new Block(pX + (pY - y), pY + (x - pX), this.color);
	}
}

/** Piece object */
function Piece(blocks) {
	this.blocks = blocks;

	/** Create a new, moved Piece. */
	this.move = function(x, y) {
		return new Piece(this.blocks.map(function(b) {
			return b.move(x, y);
		}));
	},

	/** Rotate this block clockwise around a point. */
	this.rotateClockwise = function() {
		var center = this.blocks[1];
		return new Piece(this.blocks.map(function(b) {
			return b.rotateClockwise(center.x, center.y);
		}));
	}
}

function makePiece(color, coords) {
	return new Piece(coords.map(function(c) {
		return new Block(c[0], c[1], color);
	}));
}

var line = makePiece("#00eaff", [[0, -1], [0, 0], [0, 1], [0, 2]]);
var square = makePiece("#BB0000", [[0, 0], [1, 0], [1, 1], [0, 1]]);
var l_shape1 = makePiece("#2dd400", [[-1, 0], [0, 0], [1, 0], [1, 1]]);
var l_shape2 = makePiece("#ff950c", [[-1, 0], [0, 0], [1, 0], [1, -1]]);
var n_shape1 = makePiece("#2ea4ff", [[-1, 0], [0, 0], [0, 1], [1, 1]]);
var n_shape2 = makePiece("#4b0063", [[-1, 0], [0, 0], [0, -1], [1, -1]]);
var t_shape = makePiece("#0000BB", [[-1, 0], [0, 0], [0, -1], [1, 0]]);

function randomPiece(x, y) {
	var shapes = [line, square, l_shape1, l_shape2, n_shape1, n_shape2, t_shape];
	return pickRandom(shapes).move(x, y);
}

function Board(width, height) {
	this.currentPiece = randomPiece((width / 2) - 1, 1);
	this.placedBlocks = []
	this.width = width;
	this.height = height;
	this.score = 0;

	/** Get the block at the given coordinates or null if it doesn't exist. */
	this.getBlockAt = function(x, y) {
		for ( var i = 0; i < this.placedBlocks.length; i++) {
			var b = this.placedBlocks[i];
			if (b.x == x && b.y == y) {
				return b;
			}
		}
		return null;
	};

	/** Is the given row complete? i.e. is it filled with blocks? */
	this.isRowComplete = function(y) {
		for ( var x = 0; x < this.width; x++) {
			if (this.getBlockAt(x, y) == null) {
				return false;
			}
		}
		return true;
	};

	/** Get an array of indices of completed rows. */
	this.find_completed_rows = function() {
		return range(0, this.height).filter(this.isRowComplete, this);
	};

	/**
	 * Does the given block fit on the board.
	 * i.e. is it within bounds and doesn't overlap other blocks?
	 */
	this.doesBlockFit = function(b) {
		return b.x >= 0 && b.x < this.width && b.y >= 0 && b.y < this.height
				&& this.getBlockAt(b.x, b.y) == null;
	};

	/** Does the the given piece fit on the board? */
	this.doesPieceFit = function(piece) {
		return piece.blocks.every(this.doesBlockFit, this);
	};

	/** Add the blocks of the given piece to the board */
	this.placePiece = function(piece) {
		if (!this.doesPieceFit(piece)) {
			throw "Cannot place an invalid piece";
		}
		for ( var i = 0; i < piece.blocks.length; i++) {
			this.placedBlocks.push(piece.blocks[i]);
		}
	};

	/** Remove all the blocks on the given row and shift higher blocks down. */
	this.removeRow = function(y) {
		this.placedBlocks = this.placedBlocks
			.filter(function(b) { return b.y != y;})
			.map(function(b) {
				return (b.y < y) ? b.move(0, 1) : b;
			}
		);
	}

	/** Is the given piece touching any blocks beneath it? */
	this.isPieceOnBottom = function(piece) {
		return this.doesPieceFit(piece) && !this.doesPieceFit(piece.move(0, 1));
	};

	/** Set the current piece if it fits.  Returns whether the given piece fits or not. */
	this.setCurrentPiece = function(piece) {
		var valid = this.doesPieceFit(piece);
		if (valid) {
			this.currentPiece = piece;
		}
		return valid;
	};

	/**
	 * Pushes the current piece down and removes/scores any complete rows.
	 * Returns whether the game is done or not.
	 */
	this.push_currentPiece_down = function() {
		var result = false;
		if (this.isPieceOnBottom(this.currentPiece)) {
			this.placePiece(this.currentPiece);
			var completeRows = this.find_completed_rows();
			completeRows.forEach(this.removeRow, this);
			var score_points = {
				0 : 0,
				1 : 10,
				2 : 25,
				3 : 40,
				4 : 55
			}
			this.score += score_points[completeRows.length];
			result = true;
			this.currentPiece = randomPiece((this.width / 2) - 1, 1);
		} else {
			this.currentPiece = this.currentPiece.move(0, 1);
		}
		return !this.doesPieceFit(this.currentPiece);
	};
}

File: python/blocks.py
import random

# A simple tuple of x,y,color
class Block:
    def __init__(self, x, y, color):
        self.x = x
        self.y = y
        self.color = color

    # Create a new Block, shifted down
    def down(self): return Block(self.x, self.y + 1, self.color)

    # Create a new Block, shifted left
    def left(self): return Block(self.x - 1, self.y, self.color)

    # Create a new Block, shifted right
    def right(self): return Block(self.x + 1, self.y, self.color)

    def __repr__(self): return "Block(%d, %d, %s)" % (self.x, self.y, self.color)


# Represents a logical grouping of Blocks.
# A Piece can be moved around and rotated as a group.
class Piece:
    def __init__(self, center, blocks):
        self.center = center
        self.blocks = blocks

    def _create_helper(self, func): return Piece(func(self.center), map(func, self.blocks))        

    def down(self): return self._create_helper(Block.down)

    def left(self): return self._create_helper(Block.left)

    def right(self): return self._create_helper(Block.right)

    def rotate_clockwise(self): 
        def rotator(b):
            x = self.center.x - (b.y - self.center.y)
            y = self.center.y + (b.x - self.center.x)
            return Block(x, y, self.center.color)
        return self._create_helper(rotator)

    @staticmethod
    def _create_helper2(x, y, color, coords): 
        return Piece(Block(x, y, color), [Block(a+x, b+y, color) for (a,b) in coords])

    @staticmethod
    def line(x, y): return Piece._create_helper2(x, y, "#0099FF", [[0, -1], [0, 0], [0, 1], [0, 2]])

    @staticmethod
    def square(x, y): return Piece._create_helper2(x, y, "red", [[0, 0], [1, 0], [1, 1], [0, 1]])

    @staticmethod
    def l_shape1(x, y): return Piece._create_helper2(x, y, "yellow", [[-1, 0], [0, 0], [1, 0], [1, 1]])

    @staticmethod
    def l_shape2(x, y): return Piece._create_helper2(x, y, "orange", [[-1, 0], [0, 0], [1, 0], [1, -1]])

    @staticmethod
    def n_shape1(x, y): return Piece._create_helper2(x, y, "green", [[-1, 0], [0, 0], [0, 1], [1, 1]])

    @staticmethod
    def n_shape2(x, y): return Piece._create_helper2(x, y, "brown", [[-1, 0], [0, 0], [0, -1], [1, -1]])

    # The list of all named constructors.
    @staticmethod
    def creators(): return [Piece.line, Piece.square, Piece.l_shape1, Piece.l_shape2, Piece.n_shape1, Piece.n_shape2]

    # Create a new random piece
    @staticmethod
    def random_piece(x, y): return random.choice(Piece.creators())(x, y)


class Board:

    def __init__(self, width, height):
        self._blocks = []
        self._score = 0
        self._current_piece = None
        self._is_game_done = False
        self._width = width
        self._height = height
        self._current_piece = Piece.random_piece((width / 2) - 1, 1)

    def block_at(self, x, y): return next((b for b in self._blocks if b.x == x and b.y == y), None)

    # Get the list of blocks that correspond to the given row.
    def row(self, y): return [self.block_at(x, y) for x in range(0, self._width)]

    def is_row_complete(self, y): return all([b != None for b in self.row(y)])

    # Get the list of completed rows; the indexes of each row 
    def find_complete_rows(self): return [y for y in range(0, self._height) if self.is_row_complete(y)]

    # Is the given block within bounds? 
    def is_block_valid(self, b):
        return 0 <= b.x < self._width and 0 <= b.y < self._height and self.block_at(b.x, b.y) == None

    # Is the given piece within bounds?
    def is_piece_valid(self, piece): return all([self.is_block_valid(b) for b in piece.blocks])

    def place_piece(self, piece):
        if not(self.is_piece_valid(piece)): raise ValueError("Cannot place a piece that is invalid. " + str(piece))
        for b in piece.blocks: self._blocks.append(b)
    
    def remove_row(self, y):
        for b in self.row(y): self._blocks.remove(b)
        #self.blocks = [b for b in self.blocks if b.y != y]
        self._blocks = [b.down() if b.y < y else b for b in self._blocks]

    def is_piece_on_bottom(self, piece): return not(self.is_piece_valid(piece.down()))

    def set_current_piece(self, piece):
        if (self.is_piece_valid(piece)):
            self._current_piece = piece
    def current_piece(self): return self._current_piece

    def is_game_done(self): return self._is_game_done;

    # Returns whether the piece was on the bottom after the push 
    def push_current_piece_down(self):
        result = False
        if (self.is_piece_on_bottom(self._current_piece)):
            self.place_piece(self._current_piece)
            complete_rows = self.find_complete_rows()
            for y in complete_rows: self.remove_row(y)
            score_points = {0:0, 1:10, 2:25, 3:40, 4:55}
            self._score += score_points[len(complete_rows)]
            result = True
            self._current_piece = Piece.random_piece((self._width / 2) - 1, 1)
        else:
            self._current_piece = self._current_piece.down()
        if not(self.is_piece_valid(self._current_piece)):
            print("Game is Finished");
            self._is_game_done = True
        return result
